backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/PurchaseOrderLineModularContractHandler.java

```diff
@@ -0,0 +1,149 @@
+import lombok.NonNull;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_C_OrderLine;
+import org.springframework.stereotype.Component;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Component
+public class PurchaseOrderLineModularContractHandler implements IModularContractTypeHandler<I_C_OrderLine>
+{
+	private final IOrderDAO orderDAO = Services.get(IOrderDAO.class);
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+	public final ModularContractLogDAO modularContractLogDAO;
+	public final ModularContractSettingsDAO modularContractSettingsDAO;
+
+	public PurchaseOrderLineModularContractHandler(@NonNull final ModularContractLogDAO modularContractLogDAO,
+			@NonNull final ModularContractSettingsDAO modularContractSettingsDAO)
+	{
+		this.modularContractLogDAO = modularContractLogDAO;
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+	}
+
+	@Override
+	public boolean probablyAppliesTo(@NonNull final Object model)
+	{
+		return model.getClass().isAssignableFrom(I_C_OrderLine.class);
+	}
+
+	@Override
+	@NonNull
+	public Optional<LogEntryCreateRequest> createLogEntryCreateRequest(@NonNull final I_C_OrderLine model, @NonNull final FlatrateTermId flatrateTermId)
+	{
+		final I_C_Order order = orderDAO.getById(OrderId.ofRepoId(model.getC_Order_ID()));
+
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getFor(flatrateTermId);
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
+				.stream()
+				.map(ModuleConfig::getModularContractType)
+				.filter(moduleConfig -> Objects.equals(moduleConfig.getClassName(), PurchaseOrderLineModularContractHandler.class.getName()))
+				.map(ModularContractType::getId)
+				.findFirst();
+
+		return modularContractTypeId.map(contractTypeId -> LogEntryCreateRequest.builder()
+				.contractId(flatrateTermId)
+				.productId(ProductId.ofRepoId(model.getM_Product_ID()))
+				.referencedRecord(TableRecordReference.of(I_C_OrderLine.Table_Name, model.getC_OrderLine_ID()))
+				.producerBPartnerId(BPartnerId.ofRepoId(order.getC_BPartner_ID()))
+				.invoicingBPartnerId(BPartnerId.ofRepoId(order.getBill_BPartner_ID()))
+				.collectionPointBPartnerId(BPartnerId.ofRepoId(order.getC_BPartner_ID()))
+				.collectionPoint(WarehouseId.ofRepoId(order.getM_Warehouse_ID()))
+				.documentType(LogEntryDocumentType.PURCHASE_ORDER)
+				.soTrx(SOTrx.PURCHASE)
+				.processed(false)
+				.transactionDate(InstantAndOrgId.ofTimestamp(order.getDateOrdered(), OrgId.ofRepoId(model.getAD_Org_ID())).toLocalDateAndOrgId(orgDAO::getTimeZone))
+				.year(YearId.ofRepoId(1000000))    //TODO .year(YearId.ofRepoId(order.getC_Year_ID())) replace once C_Order.C_Year_ID will be a thing
```

Comment:
is this still a TODO?

Reply:
nvm, got it (because implements IModularContractTypeHandler)

Reply:
No, Florina added the column in the meantime.